working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,431 @@
+* **Should be able to synthesize whole new classes**. Some of the core use cases
+  involve creating new classes from either other classes or even just top level
+  functions.
+* **Metaprogramming should be composable**. In one form or another,
+  metaprogramming applications should be able to be composed together. A data
+  class code generator for instance would ideally be a combination of several
+  more targeted code generators for each of the desired features (`copyWith()`,
+  `==`, `hashCode`, etc).
+* **Generated code must be debuggable and user visible**. See the Usability
+  section below.
+
+## Design constraints
+
+In addition to user-level requirements, there are some practical constraints to
+ensure that we don't compromise the existing advantages of Dart.
+
+* **Must support modular compilation**. Any individual library should be able to
+  be fully compiled (including running all macros) using only the transitive
+  dependencies of the library. To this end we should not expose any API or
+  functionality that allows global access to the program, outside of the current
+  library and/or its dependencies. Such functionality would defeat modular
+  compilation.
+* **Must support incremental compilation**. This is key to the hot reload
+  workflow—any change in the program should not have to rerun all
+  metaprogramming in the entire program. It should at worst require a rerun of
+  the metaprogramming in libraries that depend on the modified library.
+* **Must support expression evaluation**. Expression evaluation is a vital tool
+  for the development experience and we need to ensure it is not negatively
+  affected.
+
+For the most part the design constraints here align with the modular compilation
+design constraints. In short, metaprogramming should not introduce global action
+at a distance.
+
+## Non-goals
+
+* **Modifying existing code**. We don’t see a need to allow modifiying existing
+  code through metaprogramming. This leads to unexpected behavior, and
+  negatively affects usability. Dart code that already has clearly defined
+  meaning today should continue to mean the same thing after metaprogramming.
+  The goal is not to be able to hijack existing Dart syntax to do wildly
+  different things, but to augment code with new capabilities.
+
+## Approach
+
+To understand what a static metaprogramming solution might look like, we’ll walk
+through one of the use cases. Say a Dart user wants to write a library that
+automatically defines `copyWith()` methods for classes. For them to do that,
+Dart needs to provide a way for users to:
+
+1. **Invoke the metaprogramming** to control which classes should get the
+   `copyWith()` method.
+2. **Define some metaprogramming** that specifies how the copyWith() method is
+   defined in terms of the class’s fields. That metaprogramming needs to be
+   authored in some kind of metaprogramming language.
+3. It uses some sort of **introspection mechanism** to walk over the class’s
```

Comment:
Bullet 3 loses the tense.  Each of these bullets is a continuation of "provide a way for users to:", and "provide a way for users to It uses some sort of" doesn't work.  Same for bullet 4.